File: DsaProblemsInJava/src/main/java/org/example/Jan012025/Jan152025.java
package org.example.Jan012025;

import java.util.ArrayList;

public class Jan152025 {
    public int minimizeXor(int num1, int num2) {
        ArrayList<Integer> num1Arr = new ArrayList<>();
        ArrayList<Integer> num2Arr = new ArrayList<>();

        int num1c = num1;
        int num2c = num2;

        while (num1c != 0) {
            num1Arr.add(0, num1c % 2);
            num1c /= 2;
        }

        while (num2c != 0) {
            num2Arr.add(0, num2c % 2);
            num2c /= 2;
        }

        int oneCounter = 0;
        for (int elem : num2Arr) {
            if (elem == 1) {
                oneCounter++;
            }
        }

        while (num2Arr.size() > num1Arr.size()) {
            num1Arr.add(0, 0);
        }
        while (num1Arr.size() > num2Arr.size()) {
            num2Arr.add(0, 0);
        }

        ArrayList<Integer> num3Arr = new ArrayList<>();

        for (int elem : num1Arr) {
            if (elem == 0) {
                num3Arr.add(0);
            } else if (oneCounter > 0 && elem == 1) {
                num3Arr.add(1);
                oneCounter--;
            } else {
                num3Arr.add(0);
            }
        }

        if (oneCounter > 0) {
            for (int i = num3Arr.size() - 1; i >= 0; i--) {
                if (oneCounter > 0 && num3Arr.get(i) == 0) {
                    num3Arr.set(i, 1);
                    oneCounter--;
                }
            }
        }

        int factor = 1;
        int number = 0;

        for (int i = num3Arr.size() - 1; i >= 0; i--) {
            number += factor * num3Arr.get(i);
            factor *= 2;
        }

        return number;
    }
}

File: DsaProblemsInJava/src/main/java/org/example/Jan012025/PythonCodes/Jan152025.py
class Solution:
    def minimizeXor(self, num1: int, num2: int) -> int:
        num1Arr = []
        num2Arr = []
        num1c = num1
        num2c = num2
        while num1c!=0:
            num1Arr.insert(0,num1c%2)
            num1c = (num1c-num1c%2)/2

        while num2c!=0:
            num2Arr.insert(0,num2c%2)
            num2c = (num2c-num2c%2)/2


        oneCounter =0
        for elem in num2Arr:
            if elem ==1:
                oneCounter+=1
        if(len(num2Arr)>len(num1Arr)):
            while(len(num2Arr)!=len(num1Arr)):
                num1Arr.insert(0,0)
        elif(len(num1Arr)>len(num2Arr)):
            while(len(num1Arr)!=len(num2Arr)):
                num2Arr.insert(0,0)
                #print(num1Arr)
        #print(num2Arr)
        num3Arr = []
        #print(oneCounter)
        for elem in num1Arr:
            if elem==0:
                num3Arr.append(0)
            elif(oneCounter>0 and elem ==1):
                num3Arr.append(1)
                oneCounter -=1
            else:
                num3Arr.append(0)
        #print(num3Arr)
        #print()
        if(oneCounter >0):
            for i in range(len(num3Arr)-1,-1,-1):
                if(oneCounter>0 and num3Arr[i]==0):
                    num3Arr[i] =1
                    oneCounter -=1
                    #print("updated later")
        #print(num3Arr)

        factor = 1
        number = 0
        for i in range(len(num3Arr)-1,-1,-1):
            number += factor * num3Arr[i]
            factor *=2
        return number


